### ISSUE TITLE
Ambiguity isn't caught when a derived class is passed to a function template taking a `simple-template-id`

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#19208]
|  |  |
| --- | --- |
| Bugzilla Link | [18834](https://llvm.org/bz18834) |
| Version | 3.4 |
| OS | All |
| Attachments | [testcase.cpp](https://user-images.githubusercontent.com/60944935/143748990-9a1bf6e8-273e-4dbf-a5bb-643446b30908.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@zygoloid,@timsong-cpp,@jwakely |

## Extended Description 
```cpp
template<typename>
struct Base { };

struct Derived : Base<void>, Base<void*> { };

template<template<typename> class TT, typename T>
void func (TT<T>) { } 

int main () {
  func (Derived { }); 
}
```

[temp.deduct.call]/5 says that if an ambigiuity is found when trying to deduce template type parameters the code is illformed and type deduction fails.

clang incorrectly accepts the provided snippet, even though `func<Base,void>` and `func<Base,void*>` are equally possible alternatives.

Currently `func (Derived { })` will be called with `TT = Base, T = void*` (ie. the last matching base).

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [18834](https://llvm.org/bz18834) |
| Version | 3.4 |
| OS | All |
| Attachments | [testcase.cpp](https://user-images.githubusercontent.com/60944935/143748990-9a1bf6e8-273e-4dbf-a5bb-643446b30908.gz) |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@zygoloid,@timsong-cpp,@jwakely |

## Extended Description 
```cpp
template&lt;typename&gt;
struct Base { };

struct Derived : Base&lt;void&gt;, Base&lt;void*&gt; { };

template&lt;template&lt;typename&gt; class TT, typename T&gt;
void func (TT&lt;T&gt;) { } 

int main () {
  func (Derived { }); 
}
```


[temp.deduct.call]/5 says that if an ambigiuity is found when trying to deduce template type parameters the code is illformed and type deduction fails.

clang incorrectly accepts the provided snippet, even though `func&lt;Base,void&gt;` and `func&lt;Base,void*&gt;` are equally possible alternatives.

Currently `func (Derived { })` will be called with `TT = Base, T = void*` (ie. the last matching base).
</details>

**Comment by shafik** [comment]
fixed in clang-3.9.0: https://godbolt.org/z/ofafExe7K
